Port WebPreferencesStore to the new IPC serialization format

cdumez opened the pull request:
#### 87f0462fef5bff3516ba5a98f98d3506df61708f
<pre>
Port WebPreferencesStore to the new IPC serialization format
<a href="https://bugs.webkit.org/show_bug.cgi?id=264629">https://bugs.webkit.org/show_bug.cgi?id=264629</a>

Reviewed by Alex Christensen.

* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/WebPreferencesStore.cpp:
(WebKit::WebPreferencesStore::WebPreferencesStore): Deleted.
(WebKit::WebPreferencesStore::encode const): Deleted.
(WebKit::WebPreferencesStore::decode): Deleted.
* Source/WebKit/Shared/WebPreferencesStore.h:
* Source/WebKit/Shared/WebPreferencesStore.serialization.in: Added.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/270584@main">https://commits.webkit.org/270584@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/67425158f07edbadf5f07cc6cd2dbcfea22cc524

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/25821 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4428 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27100 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/27918 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23639 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6188 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/1857 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/23741 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26070 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3338 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22238 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/28498 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/2955 "Passed tests") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/29265 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/23571 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/23569 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27136 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/2968 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1196 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4359 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3424 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3313 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3284 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/Shared/WebPreferencesStore.serialization.in:
```diff
@@ -0,0 +1,28 @@
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+struct WebKit::WebPreferencesStore {
+    [NotSerialized] uint32_t ptrCount();
```
I think instead of this we want a new attribute [CanMakeCheckedPtr] that does something like VirtualTableAndRefCountOverhead but asserts that it inherits from CanMakeCheckedPtr and adds the accounting for the uint32_t.

achristensen07 commented on Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
```diff
@@ -49,6 +49,11 @@ header: <wtf/text/AtomString.h>
 }
 
+header: <wtf/CheckedRef.h>
+[CustomHeader, Nested] class WTF::CanMakeCheckedPtr {
```
Interesting.  Not technically nested, but this does the trick.  I'm not sure how I feel about this, but we can always make it marginally nicer later if we feel like it can be improved.

cdumez committed:
Port WebPreferencesStore to the new IPC serialization format
https://bugs.webkit.org/show_bug.cgi?id=264629

Reviewed by Alex Christensen.

* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/WebPreferencesStore.cpp:
(WebKit::WebPreferencesStore::WebPreferencesStore): Deleted.
(WebKit::WebPreferencesStore::encode const): Deleted.
(WebKit::WebPreferencesStore::decode): Deleted.
* Source/WebKit/Shared/WebPreferencesStore.h:
* Source/WebKit/Shared/WebPreferencesStore.serialization.in: Added.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/270584@main
